File: js/lib/experimental/PMSLocalId.ts
import { Codebooks } from "../Codebooks";
import { Gmod } from "../Gmod";
import { GmodPath } from "../GmodPath";
import { ILocalId } from "../ILocalId";
import { Locations } from "../Location";
import { MetadataTag } from "../MetadataTag";
import { VisVersion } from "../VisVersion";
import { LocalIdParsingErrorBuilder } from "../internal/LocalIdParsingErrorBuilder";
import { PMSLocalIdBuilder } from "./PMSLocalId.Builder";

export class PMSLocalId implements ILocalId<PMSLocalId> {
    private readonly _builder: PMSLocalIdBuilder;

    public get visVersion(): VisVersion {
        return this._builder.visVersion!;
    }

    public get verboseMode(): boolean {
        return this._builder.verboseMode;
    }
    public get primaryItem(): GmodPath {
        return this._builder.primaryItem!;
    }
    public get secondaryItem(): GmodPath | undefined {
        return this._builder.secondaryItem;
    }
    public get hasCustomTag(): boolean {
        return this._builder.hasCustomTag;
    }
    public get metadataTags(): MetadataTag[] {
        return this._builder.metadataTags;
    }

    constructor(builder: PMSLocalIdBuilder) {
        if (builder.isEmpty)
            throw new Error(
                "PMSLocalId cannot be constructed from empty PMSLocalIdBuilder"
            );
        if (!builder.isValid)
            throw new Error(
                "PMSLocalId cannot be constructed from invalid PMSLocalIdBuilder"
            );
        this._builder = builder;
    }

    public toString(): string {
        return this._builder.toString();
    }

    public static parse(
        localIdStr: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ): PMSLocalId {
        return PMSLocalIdBuilder.parse(
            localIdStr,
            gmod,
            codebooks,
            locations,
            errorBuilder
        ).build();
    }

    public static parseAsync(
        localIdStr: string,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return PMSLocalIdBuilder.parseAsync(localIdStr, errorBuilder);
    }
}

File: csharp/src/Vista.SDK/Experimental/PMSLocalId.cs
using Vista.SDK.Internal;

namespace Vista.SDK.Experimental;

public class PMSLocalId : ILocalId<PMSLocalId>, IEquatable<PMSLocalId>
{
    public static readonly string NamingRule = "dnv-v2";

    private readonly PMSLocalIdBuilder _builder;

    public PMSLocalIdBuilder Builder => _builder;

    internal PMSLocalId(PMSLocalIdBuilder builder)
    {
        if (builder.IsEmpty)
            throw new ArgumentException("PMSLocalId cannot be constructed from empty PMSLocalIdBuilder");
        if (!builder.IsValid)
            throw new ArgumentException("PMSLocalId cannot be constructed from invalid PMSLocalIdBuilder");
        _builder = builder;
    }

    public VisVersion VisVersion => _builder.VisVersion!.Value;

    public bool VerboseMode => _builder.VerboseMode;

    public GmodPath PrimaryItem => _builder.PrimaryItem!;

    public GmodPath? SecondaryItem => _builder.SecondaryItem;

    public MetadataTag? Quantity => _builder.Quantity;

    public MetadataTag? Content => _builder.Content;

    public MetadataTag? State => _builder.State;

    public MetadataTag? Command => _builder.Command;

    public MetadataTag? FunctionalServices => _builder.FunctionalServices;

    public MetadataTag? MaintenanceCategory => _builder.MaintenanceCategory;

    public MetadataTag? ActivityType => _builder.ActivityType;

    public MetadataTag? Position => _builder.Position;

    public MetadataTag? Detail => _builder.Detail;

    public bool HasCustomTag => _builder.HasCustomTag;

    public IReadOnlyList<MetadataTag> MetadataTags => _builder.MetadataTags;

    public override string ToString() => _builder.ToString();

    public static PMSLocalId Parse(string localIdStr) => PMSLocalIdBuilder.Parse(localIdStr).Build();

    public static bool TryParse(string localIdStr, out ParsingErrors errors, out PMSLocalId? localId)
    {
        if (!PMSLocalIdBuilder.TryParse(localIdStr, out errors, out var localIdBuilder))
        {
            localId = null;
            return false;
        }

        localId = localIdBuilder.Build();
        return true;
    }

    public sealed override bool Equals(object? obj) => Equals(obj as PMSLocalId);

    public bool Equals(PMSLocalId? other)
    {
        if (other is null)
            return false;

        return _builder.Equals(other._builder);
    }

    public static bool operator !=(PMSLocalId? left, PMSLocalId? right) => !(left == right);

    public static bool operator ==(PMSLocalId? left, PMSLocalId? right)
    {
        if (!ReferenceEquals(left, right))
        {
            if (left is not null)
                return left.Equals(right);
            return false;
        }
        return true;
    }

    public sealed override int GetHashCode() => _builder.GetHashCode();
}
